project-api/ballerina-projects/src/main/java/io/ballerina/projects/writers/BaloWriter.java

```diff
@@ -0,0 +1,390 @@
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package io.ballerina.projects.writers;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import io.ballerina.projects.Package;
+import io.ballerina.projects.model.BaloJson;
+import io.ballerina.projects.model.PackageJson;
+import io.ballerina.projects.model.adaptors.JsonCollectionsAdaptor;
+import io.ballerina.projects.model.adaptors.JsonStringsAdaptor;
+import io.ballerina.projects.utils.ProjectConstants;
+import org.ballerinalang.compiler.BLangCompilerException;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.nio.charset.Charset;
+import java.nio.file.AccessDeniedException;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystems;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.StandardCopyOption;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * {@code BaloWriter} writes a package to balo format.
+ *
+ * @since 2.0.0
+ */
+public class BaloWriter {
+
+    private BaloWriter() {}
+
+    /**
+     * Write a package to a .balo and return the created .balo path.
+     *
+     * @param pkg  Package to be written as a .balo.
+     * @param path Directory where the .balo should be created.
+     * @return Newly created balo path
+     */
+    public static Path write(Package pkg, Path path) throws AccessDeniedException {
+        // todo check if the given package is compiled properly
+
+        // Check if the path is a directory
+        if (!path.toFile().isDirectory()) {
+            throw new RuntimeException("Given path is not a directory: " + path);
+        }
+
+        if (!path.toFile().canWrite()) {
+            throw new AccessDeniedException("No write access to create balo:" + path);
+        }
+
+        Path balo = path.resolve(
+                getBaloName(pkg.packageOrg().toString(), pkg.packageName().toString(), pkg.packageVersion().toString(),
+                        null));
+
+        // Create the archive over write if exists
+        try (FileSystem baloFS = createBaloArchive(balo)) {
+            // Now lets put stuff in
+            populateBaloArchive(baloFS, pkg);
+        } catch (IOException e) {
+            throw new BLangCompilerException("Failed to create balo :" + e.getMessage(), e);
+        } catch (BLangCompilerException be) {
+            // clean up if an error occur
+            try {
+                Files.delete(balo);
+            } catch (IOException e) {
+                // We ignore this error and throw out the original blang compiler error to the user
+            }
+            throw be;
+        }
+        return balo;
+    }
+
+    private static String getBaloName(String org, String pkgName, String version, String platform) {
```

Comment:
Can we move this to util since this will be needed in other places.

Reply:
Added in 47a602a13dcd3c359d13d5fb7b731527dc77f530